MeshLib/MeshSurfaceExtraction.cpp

```diff
@@ -435,4 +394,218 @@ bool MeshSurfaceExtraction::createSfcMeshProperties(
+                    *static_cast<const MeshLib::Point*>(face.get())));
+                break;
+            }
+            default:
+                ERR("Unknown face element extracted.");
+        }
+        element_to_bulk_face_id_map.push_back(j);
+        element_to_bulk_element_id_map.push_back(surface_element.getID());
+    }
+}
+
+std::tuple<std::vector<MeshLib::Element*>, std::vector<std::size_t>,
+           std::vector<std::size_t>>
+MeshSurfaceExtraction::createSurfaceElements(MeshLib::Mesh const& bulk_mesh)
+{
+    std::vector<std::size_t> element_to_bulk_element_id_map;
+    std::vector<std::size_t> element_to_bulk_face_id_map;
+    std::vector<MeshLib::Element*> surface_elements;
+
+    auto const& bulk_elements = bulk_mesh.getElements();
+    auto const mesh_dimension = bulk_mesh.getDimension();
+
+    for (auto const* elem : bulk_elements)
+    {
+        const unsigned element_dimension(elem->getDimension());
+        if (element_dimension < mesh_dimension)
+        {
+            continue;
+        }
+
+        if (!elem->isBoundaryElement())
+        {
+            continue;
+        }
+        createSurfaceElementsFromElement(*elem, surface_elements,
+                                         element_to_bulk_element_id_map,
+                                         element_to_bulk_face_id_map);
+    }
+    return {surface_elements, element_to_bulk_element_id_map,
+            element_to_bulk_face_id_map};
+}
+
+std::unique_ptr<MeshLib::Mesh> MeshSurfaceExtraction::getBoundaryElementsAsMesh(
+    MeshLib::Mesh const& bulk_mesh,
+    std::string const& subsfc_node_id_prop_name,
+    std::string const& subsfc_element_id_prop_name,
+    std::string const& face_id_prop_name)
+{
+    auto const mesh_dimension = bulk_mesh.getDimension();
+    if (mesh_dimension < 2 || mesh_dimension > 3)
+    {
+        ERR("Cannot handle meshes of dimension %i", mesh_dimension);
+    }
+
+    // create surface elements based on the subsurface nodes
+    std::vector<std::size_t> element_to_bulk_element_id_map;
+    std::vector<std::size_t> element_to_bulk_face_id_map;
+    std::vector<MeshLib::Element*> surface_elements;
+    std::tie(surface_elements, element_to_bulk_element_id_map,
+             element_to_bulk_face_id_map) = createSurfaceElements(bulk_mesh);
+
+    // create new nodes needed for the new surface elements
+    std::vector<MeshLib::Node*> surface_nodes;
+    std::vector<std::size_t> node_id_map;
+    std::tie(surface_nodes, node_id_map) =
+        MeshSurfaceExtraction::createNodesFromElements(
+            surface_elements, bulk_mesh.getNumberOfNodes());
+
+    // create new elements using newly created nodes and delete temp-elements
+    std::vector<MeshLib::Element*> new_elements;
+    try
+    {
+        new_elements = MeshSurfaceExtraction::createSfcElementVector(
+            surface_elements, surface_nodes, node_id_map);
+    }
+    catch (std::runtime_error const& err)
```

Comment:
Who is throwing the `runtime_error`?

Reply:
The function createSfcElementVector calls OGS_FATAL that can under certain circumstances throw a `runtime_error`.